File: app/templating.py
from __future__ import annotations

from datetime import datetime, timezone

from fastapi import Request
from fastapi.templating import Jinja2Templates

from app.config import BASE_DIR, settings
from app.deps import pop_flashes

templates = Jinja2Templates(directory=str(BASE_DIR / "app" / "templates"))

STATUS_LABELS = {
    "INDEXED": "Zaindeksowany",
    "NOT_INDEXED": "Niezaindeksowany",
    "EXCLUDED": "Wykluczony",
    "UNKNOWN": "Nieznany",
    "ERROR": "Blad",
    "PENDING": "Oczekuje",
    "RUNNING": "W trakcie",
    "SUCCESS": "Sukces",
    "FAILED": "Blad",
    "SKIPPED": "Pominiety",
}

STATUS_TONES = {
    "INDEXED": "ok",
    "SUCCESS": "ok",
    "NOT_INDEXED": "warn",
    "PENDING": "warn",
    "RUNNING": "info",
    "EXCLUDED": "muted",
    "SKIPPED": "muted",
    "UNKNOWN": "muted",
    "ERROR": "bad",
    "FAILED": "bad",
}

JOB_TYPE_LABELS = {
    "URL_UPDATED": "Zgloszenie URL",
    "URL_DELETED": "Usuniecie URL",
    "INSPECT": "Inspekcja",
    "SITEMAP_SUBMIT": "Zgloszenie sitemapy",
    "SITEMAP_DELETE": "Usuniecie sitemapy",
    "INDEXNOW": "IndexNow",
}


def fmt_datetime(value: datetime | None, fmt: str = "%d.%m.%Y %H:%M") -> str:
    if value is None:
        return "—"
    return value.strftime(fmt)


def fmt_relative(value: datetime | None) -> str:
    if value is None:
        return "nigdy"
    now = datetime.now(timezone.utc).replace(tzinfo=None)
    delta = now - value
    seconds = int(delta.total_seconds())
    if seconds < 0:
        return "za chwile"
    if seconds < 60:
        return "przed chwila"
    if seconds < 3600:
        return f"{seconds // 60} min temu"
    if seconds < 86400:
        return f"{seconds // 3600} godz. temu"
    days = seconds // 86400
    if days == 1:
        return "wczoraj"
    if days < 30:
        return f"{days} dni temu"
    months = days // 30
    return f"{months} mies. temu" if months < 12 else f"{days // 365} lat temu"


def fmt_number(value) -> str:
    try:
        return f"{int(value):,}".replace(",", " ")
    except (TypeError, ValueError):
        return str(value)


def status_label(value: str | None) -> str:
    return STATUS_LABELS.get(str(value), str(value or "—"))


def status_tone(value: str | None) -> str:
    return STATUS_TONES.get(str(value), "muted")


def job_type_label(value: str | None) -> str:
    return JOB_TYPE_LABELS.get(str(value), str(value or "—"))


def truncate_url(value: str | None, length: int = 60) -> str:
    if not value:
        return "—"
    cleaned = value.replace("https://", "").replace("http://", "")
    return cleaned if len(cleaned) <= length else cleaned[: length - 1] + "…"


templates.env.filters["dt"] = fmt_datetime
templates.env.filters["ago"] = fmt_relative
templates.env.filters["num"] = fmt_number
templates.env.filters["status_label"] = status_label
templates.env.filters["status_tone"] = status_tone
templates.env.filters["job_type"] = job_type_label
templates.env.filters["short_url"] = truncate_url
templates.env.globals["app_name"] = settings.app_name
templates.env.globals["now"] = lambda: datetime.now()


def render(request: Request, template: str, context: dict | None = None, status_code: int = 200):
    payload = {
        "request": request,
        "flashes": pop_flashes(request),
        "settings": settings,
    }
    payload.update(context or {})
    return templates.TemplateResponse(request, template, payload, status_code=status_code)

File: src/templating.ts
import path from "node:path";
import type { Express, Request } from "express";
import nunjucks from "nunjucks";
import { ROOT_DIR, config } from "./config";
import { popFlashes } from "./middleware/auth";

const STATUS_LABELS: Record<string, string> = {
  INDEXED: "Zaindeksowany",
  NOT_INDEXED: "Niezaindeksowany",
  EXCLUDED: "Wykluczony",
  UNKNOWN: "Nieznany",
  ERROR: "Blad",
  PENDING: "Oczekuje",
  RUNNING: "W trakcie",
  SUCCESS: "Sukces",
  FAILED: "Blad",
  SKIPPED: "Pominiety",
};

const STATUS_TONES: Record<string, string> = {
  INDEXED: "ok",
  SUCCESS: "ok",
  NOT_INDEXED: "warn",
  PENDING: "warn",
  RUNNING: "info",
  EXCLUDED: "muted",
  SKIPPED: "muted",
  UNKNOWN: "muted",
  ERROR: "bad",
  FAILED: "bad",
};

const JOB_TYPE_LABELS: Record<string, string> = {
  URL_UPDATED: "Zgloszenie URL",
  URL_DELETED: "Usuniecie URL",
  INSPECT: "Inspekcja",
  SITEMAP_SUBMIT: "Zgloszenie sitemapy",
  SITEMAP_DELETE: "Usuniecie sitemapy",
  INDEXNOW: "IndexNow",
};

function pad(value: number): string {
  return String(value).padStart(2, "0");
}

export function fmtDatetime(value: Date | string | null | undefined): string {
  if (!value) return "—";
  const date = value instanceof Date ? value : new Date(value);
  if (Number.isNaN(date.getTime())) return "—";
  return `${pad(date.getDate())}.${pad(date.getMonth() + 1)}.${date.getFullYear()} ${pad(date.getHours())}:${pad(date.getMinutes())}`;
}

export function fmtRelative(value: Date | string | null | undefined): string {
  if (!value) return "nigdy";
  const date = value instanceof Date ? value : new Date(value);
  if (Number.isNaN(date.getTime())) return "nigdy";

  const seconds = Math.floor((Date.now() - date.getTime()) / 1000);
  if (seconds < 0) return "za chwile";
  if (seconds < 60) return "przed chwila";
  if (seconds < 3600) return `${Math.floor(seconds / 60)} min temu`;
  if (seconds < 86400) return `${Math.floor(seconds / 3600)} godz. temu`;

  const days = Math.floor(seconds / 86400);
  if (days === 1) return "wczoraj";
  if (days < 30) return `${days} dni temu`;
  const months = Math.floor(days / 30);
  return months < 12 ? `${months} mies. temu` : `${Math.floor(days / 365)} lat temu`;
}

export function fmtNumber(value: unknown): string {
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) return String(value ?? "—");
  // Spacja jako separator tysiecy, jak w polskiej konwencji.
  return Math.trunc(parsed).toString().replace(/\B(?=(\d{3})+(?!\d))/g, " ");
}

export function truncateUrl(value: string | null | undefined, length = 60): string {
  if (!value) return "—";
  const cleaned = value.replace(/^https?:\/\//, "");
  return cleaned.length <= length ? cleaned : `${cleaned.slice(0, length - 1)}…`;
}

export function configureTemplates(app: Express): nunjucks.Environment {
  const env = nunjucks.configure(path.join(ROOT_DIR, "views"), {
    autoescape: true,
    express: app,
    noCache: config.debug,
  });

  env.addFilter("dt", fmtDatetime);
  env.addFilter("ago", fmtRelative);
  env.addFilter("num", fmtNumber);
  env.addFilter("short_url", truncateUrl);
  env.addFilter("status_label", (value: unknown) => STATUS_LABELS[String(value)] ?? String(value ?? "—"));
  env.addFilter("status_tone", (value: unknown) => STATUS_TONES[String(value)] ?? "muted");
  env.addFilter("job_type", (value: unknown) => JOB_TYPE_LABELS[String(value)] ?? String(value ?? "—"));

  // Jinja2 ma |map(attribute="x")|list, Nunjucks nie - wykresy w szablonach
  // wyciagaja tym jedna kolumne z listy obiektow.
  env.addFilter("attr_list", (rows: unknown, attribute: string) => {
    if (!Array.isArray(rows)) return [];
    return rows.map((row) => (row as Record<string, unknown>)?.[attribute] ?? null);
  });

  // Odpowiednik |tojson z Jinja2. Nunjucks ma |dump, ale nie escapuje znakow,
  // ktore zamknelyby atrybut HTML, w ktorym te dane siedza.
  env.addFilter("tojson", (value: unknown) => {
    const json = JSON.stringify(value ?? null);
    return new nunjucks.runtime.SafeString(
      json.replace(/</g, "\\u003c").replace(/>/g, "\\u003e").replace(/&/g, "\\u0026").replace(/'/g, "\\u0027"),
    );
  });

  // Jinja2 ma |min/|max na listach oraz '%02d'|format(...) - w Nunjucks nie ma
  // zadnego z nich, a szablony z nich korzystaja.
  env.addFilter("min", (rows: unknown) => (Array.isArray(rows) ? Math.min(...rows.map(Number)) : rows));
  env.addFilter("max", (rows: unknown) => (Array.isArray(rows) ? Math.max(...rows.map(Number)) : rows));
  env.addFilter("pad2", (value: unknown) => pad(Number(value) || 0));
  // Jinja2 ma |round i |int; Nunjucks nie - bez nich szablony z pokryciem i ms padaja.
  env.addFilter("round", (value: unknown, precision = 0) => {
    const n = Number(value);
    if (!Number.isFinite(n)) return value;
    const factor = 10 ** Number(precision);
    return Math.round(n * factor) / factor;
  });
  env.addFilter("int", (value: unknown) => {
    const n = Number.parseInt(String(value), 10);
    return Number.isFinite(n) ? n : 0;
  });
  env.addFilter("upper", (value: unknown) => String(value ?? "").toUpperCase());

  env.addGlobal("app_name", config.appName);
  env.addGlobal("now", () => new Date());

  return env;
}

/** Adres bazy do wyswietlenia w panelu - bez hasla. */
function maskedDatabaseUrl(): string {
  const { user, host, port, database } = config.db;
  return `mysql://${user}@${host}:${port}/${database}`;
}

export interface RenderContext {
  [key: string]: unknown;
}

/** Wspolny kontekst dla kazdej strony - odpowiednik render() z wersji Pythona. */
export function baseContext(req: Request, context: RenderContext = {}): RenderContext {
  return {
    flashes: popFlashes(req),
    app_settings: {
      app_name: config.appName,
      base_url: config.baseUrl,
      host: config.host,
      port: config.port,
      database_url: maskedDatabaseUrl(),
      google_configured: config.googleConfigured,
      timezone: config.timezone,
      default_daily_quota: config.defaultDailyQuota,
      auto_index_hour: config.autoIndexHour,
      sitemap_scan_interval_hours: config.sitemapScanIntervalHours,
      recheck_after_days: config.recheckAfterDays,
      inspection_batch_size: config.inspectionBatchSize,
    },
    user: req.user ?? null,
    workspace: req.workspace ?? null,
    request: { path: req.path, query: req.query, url: req.originalUrl },
    ...context,
  };
}
